### DISCUSSION TITLE
About Uncaught Exception: 3033 listen EACCES: permission denied

### DISCUSSION THREAD

**Discussion by sangyuxiaowu** [massCodeIO/massCode#156] [Ideas]
When the Windows platform starts the software, a prompt will pop up, and the software cannot be used normally.

![image](https://user-images.githubusercontent.com/17669599/180129799-9781e1e6-fbb9-45dd-9330-95146e4c4011.png)

This problem is because windows reserves some ports, it may be because of VMware, WSL, Sandbox or other system software, not necessarily the ports are occupied.

Use the command below to see some ports reserved by windows.

```
netsh interface ipv4 show excludedportrange protocol=tcp
```

![image](https://user-images.githubusercontent.com/17669599/180130377-37b74be9-ea3a-4d99-a09e-8b8bad3d1ccc.png)

Therefore, it is recommended that the user can change the port settings monitored by the software, or check whether the port is available when the monitor is started, and try +1 to the unavailable port number one by one.

As a temporary solution, you can use the administrator command line to execute the following command.

```
net stop winnat
net start winnat
```

---------------------------------------------------

windows平台启动软件的时候会弹出提示，软件无法正常使用。

这个问题是因为 windows 保留了一些端口，可能是因为 VMware，WSL，Sandbox 或其他系统软件干的，并不一定是被占用。

使用下面的命令可以看到 windows 保留的一些端口。

```
netsh interface ipv4 show excludedportrange protocol=tcp
```

所以建议用户可以更改软件监听的端口设置，或者在启动监听时检测端口是否可用，不可用端口号逐次+1尝试。

临时的解决办法是，可以使用管理员命令行执行下面的命令

```
net stop winnat
net start winnat
```

**Comment by antonreshetov** [comment]
There is a problem in giving the user the ability to change the port on which the application runs, because on the same port also work and extensions for VS Code, Raycast & Alfred. 

I would like to provide a custom protocol, like `masscode://` but don't know how to do it properly yet. In Electron there is such a concept as [deep links](https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-app) we should dig in this direction.

**Reply by sangyuxiaowu** [reply]
> There is a problem in giving the user the ability to change the port on which the application runs, because on the same port also work and extensions for VS Code, Raycast & Alfred.
> 
> I would like to provide a custom protocol, like but don't know how to do it properly yet. In Electron there is such a concept as [deep links](https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-app) we should dig in this direction.`masscode://`

Well, the port has other uses, so it's really tricky. 
`masscode:// ` This seems to be used to evoke software for web pages, right?
like this  
![image](https://user-images.githubusercontent.com/17669599/180138258-998e9b54-45ac-4292-864b-36c8d5f4f57d.png)

**Reply by sangyuxiaowu** [reply]
This is also what I want to propose, which can expand the function of the [website](https://masscode.io/snippets/),
One-click import can be added to my code snippet using this protocol.

**Reply by antonreshetov** [reply]
I have such an idea in my note ))

**Reply by sangyuxiaowu** [reply]
> I have such an idea in my note ))

Haha, heroes see the same thing. If it wasn't for my TS basics, I would have personally helped.

**Comment by antonreshetov** [comment]
@sangyuxiaowu Yes, it looks like so - opening an application window through the browser with the transfer of something. This is how authorization works in Figma, for example.

**Reply by sangyuxiaowu** [reply]
So this does not solve the problem that the port is not available.

**Reply by antonreshetov** [reply]
Perhaps it will, if the API will also be possible to raise on this protocol. But I'm not sure.

**Reply by sangyuxiaowu** [reply]
The implementation of this protocol on Windows is to add information through the registry. Requires administrator privileges to be handled when the exe file is installed.
But the software released has no certificate and there is a windows installation warning.
So I directly unzip the exe and copy it to the folder for use.

So there may be some problems with this approach. Could try to publish service port broadcasts, but I don't know if it's available on other platforms.
